File: test/read-send-write_tests/read-write_file_r.c
#include <zmq.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <errno.h>
#include <json.h>
#include <assert.h>


int main()
{
    char *sourceFile = "/opt/HiDRA/test_file.cbf";
//    int chunkSize = 10485760; // 1024*1024*10 = 10MB
    int chunkSize = 524288;
    int filepart = 0;
    FILE *source_fp;
    void *context;
    void *socket;
    char connectionStr[128] = "ipc:///tmp/HiDRA/file_sending_test";
//    char connectionStr[128] = "tcp://zitpcx19282:55555";
    char *buffer = malloc(chunkSize);
    int bytesRead;
    int bytesSent;
    zmq_msg_t msg;
    int rc;
    json_object * metadata_json;
    const char *metadata_string;


    // Set up ZMQ
    if ( (context = zmq_ctx_new ()) == NULL )
    {
		perror("Cannot create 0MQ context");
		exit(-9);
	}

    if ( (socket = zmq_socket (context, ZMQ_PUSH)) == NULL )
    {
		perror("Could not create 0MQ dataSocket");
		exit(-9);
	}

    if ( zmq_connect(socket, connectionStr) )
    {
        fprintf(stderr, "Failed to start socket (connect) for '%s': %s\n", connectionStr, strerror( errno ));
		exit(-9);
    }
    else
    {
        printf("Socket started (connect) for '%s'\n", connectionStr);
    }

    // Open file
    source_fp = fopen(sourceFile,"rb");
    assert(source_fp != NULL);
    printf ("Opened file: %s\n", sourceFile);

    // Read file content
    while ((bytesRead = fread(buffer, 1, chunkSize, source_fp)) > 0)
    {
        printf ("Read file content of size: %i\n", bytesRead);

        // Build metadata
        metadata_json = json_object_new_object();

        json_object_object_add(metadata_json,"filename", json_object_new_string(sourceFile));
        json_object_object_add(metadata_json,"filePart", json_object_new_int(filepart));
        json_object_object_add(metadata_json,"chunkSize", json_object_new_int(chunkSize));

        metadata_string = json_object_to_json_string ( metadata_json );

        //Process message
        rc = zmq_msg_init_size (&msg, strlen(metadata_string));
        if (rc != 0) perror("ERROR when init msg");
        memcpy (zmq_msg_data (&msg), metadata_string, strlen(metadata_string));
        printf("memcpy metadata: %s\n", metadata_string);

        // Send message part
        bytesSent = zmq_msg_send (&msg, socket, ZMQ_SNDMORE);
        if (bytesSent == -1) perror("Sending message failed\n");
        printf("Sent message of size: %d\n", bytesSent);

        rc = zmq_msg_close(&msg);
        if (rc != 0) perror("ERROR when closing msg");
        printf("close message\n");

        //Process message
        rc = zmq_msg_init_size (&msg, bytesRead);
        if (rc != 0) perror("ERROR when init msg");
        memcpy (zmq_msg_data (&msg), buffer, bytesRead);
        printf("bytesRead: %i\n", bytesRead);
//        printf("memcpy data: %s\n", (char *) zmq_msg_data (&msg));

        /* Final part; no more parts to follow */
        bytesSent = zmq_msg_send (&msg, socket, 0);
        if (bytesSent == -1) perror("Sending message failed\n");
        printf("Sent message of size: %d\n", bytesSent);

        rc = zmq_msg_close(&msg);
        if (rc != 0) perror("ERROR when closing msg");

        filepart++;
        json_object_put ( metadata_json );
    }

    free(buffer);


    // Close file
    fclose(source_fp);
    printf ("Closed file: %s\n", sourceFile);

    // Clean up ZMQ
    zmq_close(socket);
    zmq_ctx_destroy(context);

    return 0;
}

File: test/coding_tests/read-send-write_tests/read-write_file_r.py
from __future__ import print_function

import os
import json
import tempfile
import zmq

try:
    CURRENT_DIR = os.path.dirname(os.path.realpath(__file__))
except NameError:
    CURRENT_DIR = os.path.dirname(os.path.abspath(sys.argv[0]))
BASE_DIR = os.path.dirname(os.path.dirname(CURRENT_DIR))
print(BASE_DIR)


def main():

    source_file = os.path.join(BASE_DIR, "test", "test_files", "test_file.cbf")
    chunksize = 10485760  # 1024*1024*10 = 10MB
    filepart = 0
    connection_str = "ipc://{}".format(
        os.path.join(tempfile.gettempdir(), "hidra", "file_sending_test")
    )

    # Set up ZMQ
    context = zmq.Context()
    socket = context.socket(zmq.PUSH)
    socket.connect(connection_str)
    print("Socket started (connect) for", connection_str)

    # Set up metadata
    metadata = {
            "filename": "test.cbf",
            "filePart": 0,
            "chunkSize": chunksize
            }

    # Open file
    source_fp = open(source_file, "rb")
    print("Opened file:", source_file)

    while True:
        # Read file content
        content = source_fp.read(chunksize)
        print("Read file content")

        if not content:
            print("break")
            break

        # Send content over ZMQ
        metadata["filePart"] = filepart

        payload = [json.dumps(metadata), content]

        tracker = socket.send_multipart(payload, copy=False, track=True)
        if not tracker.done:
            tracker.wait(2)
        if not tracker.done:
            print("Failed to send ALIVE_TEST")
        else:
            print("Sending multipart message...success")

        filepart += 1

    # Close file
    source_fp.close()
    print("Closed file:", source_file)

    # Clean up ZMQ
    socket.close(0)
    context.destroy(0)


if __name__ == '__main__':
    main()

